### ISSUE TITLE
[JENKINS-42519] Closed pull requests, branch jobs are not cleaned up

### ISSUE THREAD

**Issue by jenkins-infra-bot** [jenkinsci/github-branch-source-plugin#1061]
<p>When I am building a multibranch pipeline job, I am noticing that a lot of jobs in Jenkins are still showing and that their workspaces can still be found on disk (wasting space). It seems like when a github webhook event is received that a PR is closed or a branch is deleted, then those should be cleaned up automatically.</p>

<p> </p>

<p>Steps:</p>
<ol>
	<li>Create multi branch pipeline.</li>
	<li>Set branch source to GitHub repo.</li>
	<li>Configure to build all branches and PRs.</li>
	<li>Create PR, confirm that job runs successfully, etc.</li>
	<li>Merge PR.<br/>
 <del>=</del>=- From "Scan Repository > Events"<br/>
 <span class="error">&#91;Mon Mar 06 19:49:54 UTC 2017&#93;</span> Received org.jenkinsci.plugins.github_branch_source.PushGHEventSubscriber$SCMHeadEventImpl REMOVED event from 10.50.226.16 ⇒ https://<jenkins-host>/github-webhook/ with timestamp Mon Mar 06 19:49:49 UTC 2017<br/>
 <del>=</del>=-</li>
</ol>

<p>Results:</p>
<ul class="alternate" type="square">
	<li>PR job shows as disabled (strikethrough) in Jenkins</li>
	<li>Workspace is still on disk</li>
</ul>

<p>Expected:</p>
<ul class="alternate" type="square">
	<li>PR job is no longer shown in Jenkins</li>
	<li>Workspace is cleaned up.</li>
</ul>

<p>A similar behavior is also seen when a branch is deleted outside of a PR.</p>

<p>Workarounds:<br/>
 If I manually trigger "Scan Repository > Run Now", I see the expected results. There may also be a workaround to have scan run automatically (scheduled), but I'm not sure how to do that yet.</p>

<p> </p>

---
<details><summary><i>Originally reported by <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=stephenash">stephenash</a>, imported from: <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-42519" target="_blank">Closed pull requests, branch jobs are not cleaned up</a></i></summary>
<i><ul>
<li><b>assignee</b>: <img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/stephenconnolly.png" title="stephenconnolly's avatar" /> <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=stephenconnolly">stephenconnolly</a>
<li><b>status</b>: Resolved
<li><b>priority</b>: Minor
<li><b>component(s)</b>: github-branch-source-plugin
<li><b>resolution</b>: Not A Defect
<li><b>resolved</b>: 2018-03-15T21:43:43+00:00
<li><b>votes</b>: 0
<li><b>watchers</b>: 15
<li><b>imported</b>: 2025-12-02
</ul></i>
<details><summary>Raw content of original issue</summary>

<pre>
<p>When I am building a multibranch pipeline job, I am noticing that a lot of jobs in Jenkins are still showing and that their workspaces can still be found on disk (wasting space). It seems like when a github webhook event is received that a PR is closed or a branch is deleted, then those should be cleaned up automatically.</p>

<p> </p>

<p>Steps:</p>
<ol>
	<li>Create multi branch pipeline.</li>
	<li>Set branch source to GitHub repo.</li>
	<li>Configure to build all branches and PRs.</li>
	<li>Create PR, confirm that job runs successfully, etc.</li>
	<li>Merge PR.
 <del>=</del>=- From "Scan Repository &gt; Events"
 <span class="error">&#91;Mon Mar 06 19:49:54 UTC 2017&#93;</span> Received org.jenkinsci.plugins.github_branch_source.PushGHEventSubscriber$SCMHeadEventImpl REMOVED event from 10.50.226.16 ⇒ https://&lt;jenkins-host&gt;/github-webhook/ with timestamp Mon Mar 06 19:49:49 UTC 2017
 <del>=</del>=-</li>
</ol>

<p>Results:</p>
<ul class="alternate" type="square">
	<li>PR job shows as disabled (strikethrough) in Jenkins</li>
	<li>Workspace is still on disk</li>
</ul>

<p>Expected:</p>
<ul class="alternate" type="square">
	<li>PR job is no longer shown in Jenkins</li>
	<li>Workspace is cleaned up.</li>
</ul>

<p>A similar behavior is also seen when a branch is deleted outside of a PR.</p>

<p>Workarounds:
 If I manually trigger "Scan Repository &gt; Run Now", I see the expected results. There may also be a workaround to have scan run automatically (scheduled), but I'm not sure how to do that yet.</p>

<p> </p></pre>
</details>
</details>
<details><summary><i>environment</i></summary>

```
- Jenkins 2.32.2 LTS<br/>
- Branch API Plugin 2.0.7<br/>
- Git plugin 3.1.0<br/>
- GitHub Branch Source Plugin 2.0.3<br/>
- Pipeline: Multibranch 2.13
```
</details>
<details><summary><i>1 attachment</i></summary>

- [image-2019-11-27-08-19-32-533.png](https://raw.githubusercontent.com/jenkinsci/attachments-from-jira-issues-last/refs/heads/main/attachments/49609/image-2019-11-27-08-19-32-533.png)
  > ![image-2019-11-27-08-19-32-533.png](https://raw.githubusercontent.com/jenkinsci/attachments-from-jira-issues-last/refs/heads/main/attachments/49609/image-2019-11-27-08-19-32-533.png)
</details>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/recampbell.png" title="recampbell's avatar" /> recampbell:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-42519?focusedId=292068&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-292068">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>What is your orphaned item strategy set to?</p>
</pre>
</details>

</ul>
</details>

<p>What is your orphaned item strategy set to?</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>stephenash:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-42519?focusedId=292071&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-292071">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>"Discard old items" is checked.</p>

<p>Days to keep old items = 0</p>

<p>Max # of old items to keep = 0</p>

<p> </p>

<p>If I have "Periodically if not otherwise run" set to some value it will also clean things up, but that was trying to avoid using that because of issues described in <a href="https://issues.jenkins.io/browse/JENKINS-37491" title="Build Origin PRs (merge with base branch) conducts rebuilds when baseline changes" class="issue-link" data-issue-key="JENKINS-37491"><del>JENKINS-37491</del></a>.</p>
</pre>
</details>

</ul>
</details>

<p>"Discard old items" is checked.</p>

<p>Days to keep old items = 0</p>

<p>Max # of old items to keep = 0</p>

<p> </p>

<p>If I have "Periodically if not otherwise run" set to some value it will also clean things up, but that was trying to avoid using that because of issues described in <a href="https://issues.jenkins.io/browse/JENKINS-37491" title="Build Origin PRs (merge with base branch) conducts rebuilds when baseline changes" class="issue-link" data-issue-key="JENKINS-37491"><del>JENKINS-37491</del></a>.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/stephenconnolly.png" title="stephenconnolly's avatar" /> stephenconnolly:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-42519?focusedId=292118&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-292118">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>You need to run periodically, if even once a week</p>
</pre>
</details>

</ul>
</details>

<p>You need to run periodically, if even once a week</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>stephenash:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-42519?focusedId=292163&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-292163">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<blockquote><p>You need to run periodically, if even once a week</p></blockquote>

<p>Maybe I'm confused about the role of branch indexing. Since Jenkins got the webhook about the merged PR (or a deleted branch), why wouldn't the related jobs and workspaces be removed right then? Deferring that action until branch indexing runs (and that may not run if "Periodically if not otherwise run" is unchecked by the user) seems wrong to me.</p>
</pre>
</details>

</ul>
</details>

<blockquote><p>You need to run periodically, if even once a week</p></blockquote>

<p>Maybe I'm confused about the role of branch indexing. Since Jenkins got the webhook about the merged PR (or a deleted branch), why wouldn't the related jobs and workspaces be removed right then? Deferring that action until branch indexing runs (and that may not run if "Periodically if not otherwise run" is unchecked by the user) seems wrong to me.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/stephenconnolly.png" title="stephenconnolly's avatar" /> stephenconnolly:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-42519?focusedId=292165&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-292165">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Well if your orphaned item strategy says keep nothing, you might not have that issue... but when the orphaned item strategy says keep the last 3 branches...</p>
<ul>
	<li>There are two branches that are already orphaned.</li>
	<li>You get three events for branch deletions.</li>
	<li>The events are processing in parallel.</li>
	<li>How does each event magically know that there are two other events in flight and hence the two existing orphaned branches are now pushed over the edge and need to be deleted?</li>
</ul>


<p>We may look into augmenting the orphaned branch strategy to allow it to be run post-event, but for now the reality is that the only way it can correctly run is when it knows a consistent complete list of all live branches (and consequently has a consistent complete set of all dead branches)</p>
</pre>
</details>

</ul>
</details>

<p>Well if your orphaned item strategy says keep nothing, you might not have that issue... but when the orphaned item strategy says keep the last 3 branches...</p>
<ul>
	<li>There are two branches that are already orphaned.</li>
	<li>You get three events for branch deletions.</li>
	<li>The events are processing in parallel.</li>
	<li>How does each event magically know that there are two other events in flight and hence the two existing orphaned branches are now pushed over the edge and need to be deleted?</li>
</ul>


<p>We may look into augmenting the orphaned branch strategy to allow it to be run post-event, but for now the reality is that the only way it can correctly run is when it knows a consistent complete list of all live branches (and consequently has a consistent complete set of all dead branches)</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/stephenconnolly.png" title="stephenconnolly's avatar" /> stephenconnolly:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-42519?focusedId=292166&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-292166">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Also, consider the use case of users wanting to trigger a rebuild of a PR.</p>

<p>The way that happens right now is you close the PR and open it again.</p>

<p>If we deleted the branch immediately on the close event then the PR build history would be destroyed so that when re-opened it would start from build 1</p>
</pre>
</details>

</ul>
</details>

<p>Also, consider the use case of users wanting to trigger a rebuild of a PR.</p>

<p>The way that happens right now is you close the PR and open it again.</p>

<p>If we deleted the branch immediately on the close event then the PR build history would be destroyed so that when re-opened it would start from build 1</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/allan_burdajewicz.png" title="allan_burdajewicz's avatar" /> allan_burdajewicz:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-42519?focusedId=296174&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-296174">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>So just to confirm, is this the expected behavior to mark the branch / PR as dead and deleted them through the Periodic Branch Indexing ? Even though the Orphaned Item Strategy is set to <em>Days to keep old items = 0</em> and <em>Max # of old items to keep = 0</em> ?</p>
</pre>
</details>

</ul>
</details>

<p>So just to confirm, is this the expected behavior to mark the branch / PR as dead and deleted them through the Periodic Branch Indexing ? Even though the Orphaned Item Strategy is set to <em>Days to keep old items = 0</em> and <em>Max # of old items to keep = 0</em> ?</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/stephenconnolly.png" title="stephenconnolly's avatar" /> stephenconnolly:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-42519?focusedId=296230&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-296230">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>The current expected behaviour is that orphaned items are only removed during a full index/scan (and only if the strategy does not veto their removal)</p>
</pre>
</details>

</ul>
</details>

<p>The current expected behaviour is that orphaned items are only removed during a full index/scan (and only if the strategy does not veto their removal)</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/stephenconnolly.png" title="stephenconnolly's avatar" /> stephenconnolly:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-42519?focusedId=296232&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-296232">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>From my understanding this is functioning as designed</p>
</pre>
</details>

</ul>
</details>

<p>From my understanding this is functioning as designed</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>acejam:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-42519?focusedId=332206&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-332206">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>I would like to bump this back up. Orphaned jobs should be deleted when an incoming webhook is received if the item strategy is set to 0. I also agree with <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=stephenash" class="user-hover" rel="stephenash">stephenash</a> about <a href="https://issues.jenkins.io/browse/JENKINS-37491" title="Build Origin PRs (merge with base branch) conducts rebuilds when baseline changes" class="issue-link" data-issue-key="JENKINS-37491"><del>JENKINS-37491</del></a>. If a user wants to use GitHub Organization folders via the branch source plugin and build "merged" PR's, they either have to deal with build storms (by having "Periodically if not otherwise run" turned on) or deal with orphaned jobs. (By having "Periodically if not otherwise run" turned off, to prevent build storms)</p>
</pre>
</details>

</ul>
</details>

<p>I would like to bump this back up. Orphaned jobs should be deleted when an incoming webhook is received if the item strategy is set to 0. I also agree with <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=stephenash" class="user-hover" rel="stephenash">stephenash</a> about <a href="https://issues.jenkins.io/browse/JENKINS-37491" title="Build Origin PRs (merge with base branch) conducts rebuilds when baseline changes" class="issue-link" data-issue-key="JENKINS-37491"><del>JENKINS-37491</del></a>. If a user wants to use GitHub Organization folders via the branch source plugin and build "merged" PR's, they either have to deal with build storms (by having "Periodically if not otherwise run" turned on) or deal with orphaned jobs. (By having "Periodically if not otherwise run" turned off, to prevent build storms)</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/stephenconnolly.png" title="stephenconnolly's avatar" /> stephenconnolly:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-42519?focusedId=332270&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-332270">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>It is not possible with the design of the API to know if the event means that the job should be deleted or if a lower priority source in the multibranch project has already received an event for the same branch being created because the branch is being moved from one source to another.</p>

<p>e.g.</p>

<p>if you configure two sources in the same multibranch:</p>

<ul>
	<li>git@git.example.com:/development.git</li>
</ul>


<ul>
	<li>git@git.example.com:/production.git</li>
</ul>


<p>Now suppose you do something like:</p>
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">$ git remote add origin git@git.example.com:/development.git
$ git remote add production git@git.example.com:/production.git
$ git fetch origin feature-1
$ git push production feature-1
$ git push origin :feature-1</pre>
</div></div>
<p>So what happens if we follow <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=acejam" class="user-hover" rel="acejam">acejam</a>'s suggestion is</p>
<ol>
	<li>Jenkins gets the event from "production" saying "new branch 'feature-1'" and we say: "cool, but development is higher in the list of sources and already owns that branch name... I'll just drop that event on the floor"</li>
	<li>Jenkins gets the event from "development" saying "branch 'feature-1' has been deleted" and we'd delete the job and its history.</li>
</ol>


<p>So you would lose all the job history of feature-1 being developed.</p>

<p>What happens currently is as follows:</p>
<ol>
	<li>Jenkins gets the event from "production" saying "new branch 'feature-1'" and we say: "cool, but development is higher in the list of sources and already owns that branch name... I'll just drop that event on the floor"</li>
	<li>Jenkins gets the event from "development" saying "branch 'feature-1' has been deleted" and we mark the branch as dead.</li>
	<li>If there is a subsequent event from "production" we let that event "take over" the branch, and all is good, history retained.</li>
	<li>When the scan comes along, we get the full list of branches from each source in priority order and we assign the branch to "production" as a live branch, because the revision is the same, we don't trigger a rebuild</li>
</ol>

</pre>
</details>

</ul>
</details>

<p>It is not possible with the design of the API to know if the event means that the job should be deleted or if a lower priority source in the multibranch project has already received an event for the same branch being created because the branch is being moved from one source to another.</p>

<p>e.g.</p>

<p>if you configure two sources in the same multibranch:</p>

<ul>
	<li>git@​git.example.com:/development.git</li>
</ul>


<ul>
	<li>git@​git.example.com:/production.git</li>
</ul>


<p>Now suppose you do something like:</p>

<pre>
$ git remote add origin git@​git.example.com:/development.git
$ git remote add production git@​git.example.com:/production.git
$ git fetch origin feature-1
$ git push production feature-1
$ git push origin :feature-1</pre>
<p>So what happens if we follow <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=acejam" class="user-hover" rel="acejam">acejam</a>'s suggestion is</p>
<ol>
	<li>Jenkins gets the event from "production" saying "new branch 'feature-1'" and we say: "cool, but development is higher in the list of sources and already owns that branch name... I'll just drop that event on the floor"</li>
	<li>Jenkins gets the event from "development" saying "branch 'feature-1' has been deleted" and we'd delete the job and its history.</li>
</ol>


<p>So you would lose all the job history of feature-1 being developed.</p>

<p>What happens currently is as follows:</p>
<ol>
	<li>Jenkins gets the event from "production" saying "new branch 'feature-1'" and we say: "cool, but development is higher in the list of sources and already owns that branch name... I'll just drop that event on the floor"</li>
	<li>Jenkins gets the event from "development" saying "branch 'feature-1' has been deleted" and we mark the branch as dead.</li>
	<li>If there is a subsequent event from "production" we let that event "take over" the branch, and all is good, history retained.</li>
	<li>When the scan comes along, we get the full list of branches from each source in priority order and we assign the branch to "production" as a live branch, because the revision is the same, we don't trigger a rebuild</li>
</ol>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/stephenconnolly.png" title="stephenconnolly's avatar" /> stephenconnolly:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-42519?focusedId=332271&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-332271">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Now if we have the order different:</p>
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">$ git remote add origin git@git.example.com:/development.git
$ git remote add production git@git.example.com:/production.git
$ git fetch origin feature-1
$ git push origin :feature-1
$ git push production feature-1
</pre>
</div></div>
<p>In this case, it is even more critical that we do not delete the branch immediately as if we followed <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=acejam" class="user-hover" rel="acejam">acejam</a>'s suggestion the job history would be erased by the origin branch being deleted and then the production branch would be starting from build #1 again.</p>

<p> </p>

<p>If the user actually wants the build history reset, then all they need to do is use the Jenkins UI to delete the dead branch before doing the push of the branch to production</p>
</pre>
</details>

</ul>
</details>

<p>Now if we have the order different:</p>

<pre>
$ git remote add origin git@​git.example.com:/development.git
$ git remote add production git@​git.example.com:/production.git
$ git fetch origin feature-1
$ git push origin :feature-1
$ git push production feature-1
</pre>
<p>In this case, it is even more critical that we do not delete the branch immediately as if we followed <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=acejam" class="user-hover" rel="acejam">acejam</a>'s suggestion the job history would be erased by the origin branch being deleted and then the production branch would be starting from build #1 again.</p>

<p> </p>

<p>If the user actually wants the build history reset, then all they need to do is use the Jenkins UI to delete the dead branch before doing the push of the branch to production</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>acejam:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-42519?focusedId=332336&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-332336">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Isn't the whole point of a multibranch pipeline (within the context of the GitHub Branch Source plugin) to build multiple branches of the same origin? I don't see how adding multiple origins is a valid example.</p>
</pre>
</details>

</ul>
</details>

<p>Isn't the whole point of a multibranch pipeline (within the context of the GitHub Branch Source plugin) to build multiple branches of the same origin? I don't see how adding multiple origins is a valid example.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/stephenconnolly.png" title="stephenconnolly's avatar" /> stephenconnolly:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-42519?focusedId=332368&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-332368">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>The original design of the API was that pull requests would be handled as a separate source in the same multibranch... similarly you can add forks of the same repo... now in retrospect it one could argue that it was a bad design decision to do it that way in 2013... but there are users who use the multiple sources, so we cannot remove it... which means a branch being deleted in one source does not mean the branch job should be deleted.</p>
</pre>
</details>

</ul>
</details>

<p>The original design of the API was that pull requests would be handled as a separate source in the same multibranch... similarly you can add forks of the same repo... now in retrospect it one could argue that it was a bad design decision to do it that way in 2013... but there are users who use the multiple sources, so we cannot remove it... which means a branch being deleted in one source does not mean the branch job should be deleted.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/anton_yurchenko.png" title="anton_yurchenko's avatar" /> anton_yurchenko:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-42519?focusedId=380610&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-380610">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>We still encounter that issue on Jenkins 2.190.3.</p>

<p>Is there a proper workaround?</p>

<p><span class="image-wrap" style=""><img src="https://issues.jenkins.io/secure/attachment/49609/49609_image-2019-11-27-08-19-32-533.png" height="174" width="1190" style="border: 0px solid black" /></span></p>

<p> </p>
<div class="preformatted panel" style="border-width: 1px;"><div class="preformattedContent panelContent">
<pre>/var/lib/jenkins/jobs/Projects/jobs/&lt;project&gt;/branches&gt; $ du -h --max-depth=1 | grep "PR-"
11M	./PR-406
21M	./PR-403
11M	./PR-404
11M	./PR-405</pre>
</div></div>
<p>(those sizes are small as we dynamically provision build nodes and nothing gets executed on master. much more space consumed on actual build nodes.)</p>

<p> </p>

<p>"<b>Scan Repository Now</b>" cleans those PR's.</p>

<p>"<b>Scan Repository Log</b>":</p>
<div class="preformatted panel" style="border-width: 1px;"><div class="preformattedContent panelContent">
<pre>Evaluating orphaned items in Project » &lt;project&gt;
Will remove PR-406
Will remove PR-405
Will remove PR-404
Will remove PR-403
Finished: SUCCESS</pre>
</div></div>
<p>Relevant logs from <b>/var/log/jenkins/jenkins.log</b>:</p>
<div class="preformatted panel" style="border-width: 1px;"><div class="preformattedContent panelContent">
<pre>2019-11-27 06:28:59.389+0000 [id=234]   INFO    j.b.WorkspaceLocatorImpl$Deleter$CleanupTask#run: deleting obsolete workspace /home/user/workspace/ct_&lt;project&gt;_PR-403@tmp on &lt;slave-name&gt;
2019-11-27 06:28:59.468+0000 [id=233]   INFO    j.b.WorkspaceLocatorImpl$Deleter$CleanupTask#run: deleting obsolete workspace /var/lib/jenkins/workspace/ct_&lt;project&gt;_PR-403@libs on master
2019-11-27 06:28:59.470+0000 [id=256]   INFO    j.b.WorkspaceLocatorImpl$Deleter$CleanupTask#run: deleting obsolete workspace /var/lib/jenkins/workspace/ct_&lt;project&gt;_PR-404@libs on master
2019-11-27 06:28:59.578+0000 [id=249]   INFO    j.b.WorkspaceLocatorImpl$Deleter$CleanupTask#run: deleting obsolete workspace /home/user/workspace/ct_&lt;project&gt;_PR-404@tmp on &lt;slave-name&gt;
2019-11-27 06:28:59.668+0000 [id=247]   INFO    j.b.WorkspaceLocatorImpl$Deleter$CleanupTask#run: deleting obsolete workspace /var/lib/jenkins/workspace/ct_&lt;project&gt;_PR-405@libs on master
2019-11-27 06:28:59.927+0000 [id=264]   INFO    j.b.WorkspaceLocatorImpl$Deleter$CleanupTask#run: deleting obsolete workspace /var/lib/jenkins/workspace/ct_&lt;project&gt;_PR-406@libs on master
2019-11-27 06:29:00.084+0000 [id=239]   INFO    j.b.WorkspaceLocatorImpl$Deleter$CleanupTask#run: deleting obsolete workspace /home/user/workspace/ct_&lt;project&gt;_PR-406@tmp on &lt;slave-name&gt;
2019-11-27 06:29:00.145+0000 [id=252]   INFO    j.b.WorkspaceLocatorImpl$Deleter$CleanupTask#run: deleting obsolete workspace /home/user/workspace/ct_&lt;project&gt;_PR-405@tmp on &lt;slave-name&gt;
2019-11-27 06:29:00.638+0000 [id=233]   INFO    j.b.WorkspaceLocatorImpl$Deleter$CleanupTask#run: deleting obsolete workspace /var/lib/jenkins/workspace/&lt;omitted&gt;@libs on master
2019-11-27 06:29:00.780+0000 [id=239]   INFO    j.b.WorkspaceLocatorImpl$Deleter$CleanupTask#run: deleting obsolete workspace /home/user/workspace/&lt;omitted&gt;@tmp on &lt;slave-name&gt;
2019-11-27 06:29:01.113+0000 [id=239]   INFO    j.b.WorkspaceLocatorImpl$Deleter$CleanupTask#run: deleting obsolete workspace /var/lib/jenkins/workspace/&lt;omitted&gt;@libs on master
2019-11-27 06:29:01.154+0000 [id=229]   INFO    j.b.MultiBranchProject$BranchIndexing#run: Project/&lt;project&gt; #20191127.062845 branch indexing action completed: SUCCESS in 15 sec
2019-11-27 06:29:01.259+0000 [id=264]   INFO    j.b.WorkspaceLocatorImpl$Deleter$CleanupTask#run: deleting obsolete workspace /home/user/workspace/&lt;omitted&gt;@tmp on &lt;slave-name&gt;</pre>
</div></div>
<p> Space consumed on master before re-scanning all projects:</p>
<div class="preformatted panel" style="border-width: 1px;"><div class="preformattedContent panelContent">
<pre>Filesystem      Size  Used Avail Use% Mounted on
/dev/nvme1n1p1  150G   86G   65G  58% /var/lib/jenkins</pre>
</div></div>
<p>Space consumed on master after re-scanning all projects:</p>
<div class="preformatted panel" style="border-width: 1px;"><div class="preformattedContent panelContent">
<pre>Filesystem      Size  Used Avail Use% Mounted on
/dev/nvme1n1p1  150G   68G   83G  45% /var/lib/jenkins</pre>
</div></div>
<p>You may clearly see 20GB of junk in just 2 month on master node, when nothing gets build on it. All our pipelines contains:</p>
<div class="preformatted panel" style="border-width: 1px;"><div class="preformattedContent panelContent">
<pre>pipeline {
    agent none
    [...]    
Stages {
        [...]
        Stage('Build) {
            agent { label 'slave' }
            [...]
        }
        [...]
    }
}</pre>
</div></div>
<p>Much more space wasted on build nodes.</p>
</pre>
</details>

</ul>
</details>

<p>We still encounter that issue on Jenkins 2.190.3.</p>

<p>Is there a proper workaround?</p>

<p><span class="image-wrap" style=""><img src="https://raw.githubusercontent.com/jenkinsci/attachments-from-jira-issues-last/refs/heads/main/attachments/49609/image-2019-11-27-08-19-32-533.png" height="174" width="1190" style="border: 0px solid black" /></span></p>

<p> </p>


```
/var/lib/jenkins/jobs/Projects/jobs/<project>/branches> $ du -h --max-depth=1 | grep "PR-"
11M	./PR-406
21M	./PR-403
11M	./PR-404
11M	./PR-405
```
<p>(those sizes are small as we dynamically provision build nodes and nothing gets executed on master. much more space consumed on actual build nodes.)</p>

<p> </p>

<p>"<b>Scan Repository Now</b>" cleans those PR's.</p>

<p>"<b>Scan Repository Log</b>":</p>


```
Evaluating orphaned items in Project » <project>
Will remove PR-406
Will remove PR-405
Will remove PR-404
Will remove PR-403
Finished: SUCCESS
```
<p>Relevant logs from <b>/var/log/jenkins/jenkins.log</b>:</p>


```
2019-11-27 06:28:59.389+0000 [id=234]   INFO    j.b.WorkspaceLocatorImpl$Deleter$CleanupTask#run: deleting obsolete workspace /home/user/workspace/ct_<project>_PR-403@​tmp on <slave-name>
2019-11-27 06:28:59.468+0000 [id=233]   INFO    j.b.WorkspaceLocatorImpl$Deleter$CleanupTask#run: deleting obsolete workspace /var/lib/jenkins/workspace/ct_<project>_PR-403@​libs on master
2019-11-27 06:28:59.470+0000 [id=256]   INFO    j.b.WorkspaceLocatorImpl$Deleter$CleanupTask#run: deleting obsolete workspace /var/lib/jenkins/workspace/ct_<project>_PR-404@​libs on master
2019-11-27 06:28:59.578+0000 [id=249]   INFO    j.b.WorkspaceLocatorImpl$Deleter$CleanupTask#run: deleting obsolete workspace /home/user/workspace/ct_<project>_PR-404@​tmp on <slave-name>
2019-11-27 06:28:59.668+0000 [id=247]   INFO    j.b.WorkspaceLocatorImpl$Deleter$CleanupTask#run: deleting obsolete workspace /var/lib/jenkins/workspace/ct_<project>_PR-405@​libs on master
2019-11-27 06:28:59.927+0000 [id=264]   INFO    j.b.WorkspaceLocatorImpl$Deleter$CleanupTask#run: deleting obsolete workspace /var/lib/jenkins/workspace/ct_<project>_PR-406@​libs on master
2019-11-27 06:29:00.084+0000 [id=239]   INFO    j.b.WorkspaceLocatorImpl$Deleter$CleanupTask#run: deleting obsolete workspace /home/user/workspace/ct_<project>_PR-406@​tmp on <slave-name>
2019-11-27 06:29:00.145+0000 [id=252]   INFO    j.b.WorkspaceLocatorImpl$Deleter$CleanupTask#run: deleting obsolete workspace /home/user/workspace/ct_<project>_PR-405@​tmp on <slave-name>
2019-11-27 06:29:00.638+0000 [id=233]   INFO    j.b.WorkspaceLocatorImpl$Deleter$CleanupTask#run: deleting obsolete workspace /var/lib/jenkins/workspace/<omitted>@​libs on master
2019-11-27 06:29:00.780+0000 [id=239]   INFO    j.b.WorkspaceLocatorImpl$Deleter$CleanupTask#run: deleting obsolete workspace /home/user/workspace/<omitted>@​tmp on <slave-name>
2019-11-27 06:29:01.113+0000 [id=239]   INFO    j.b.WorkspaceLocatorImpl$Deleter$CleanupTask#run: deleting obsolete workspace /var/lib/jenkins/workspace/<omitted>@​libs on master
2019-11-27 06:29:01.154+0000 [id=229]   INFO    j.b.MultiBranchProject$BranchIndexing#run: Project/<project> #20191127.062845 branch indexing action completed: SUCCESS in 15 sec
2019-11-27 06:29:01.259+0000 [id=264]   INFO    j.b.WorkspaceLocatorImpl$Deleter$CleanupTask#run: deleting obsolete workspace /home/user/workspace/<omitted>@​tmp on <slave-name>
```
<p> Space consumed on master before re-scanning all projects:</p>


```
Filesystem      Size  Used Avail Use% Mounted on
/dev/nvme1n1p1  150G   86G   65G  58% /var/lib/jenkins
```
<p>Space consumed on master after re-scanning all projects:</p>


```
Filesystem      Size  Used Avail Use% Mounted on
/dev/nvme1n1p1  150G   68G   83G  45% /var/lib/jenkins
```
<p>You may clearly see 20GB of junk in just 2 month on master node, when nothing gets build on it. All our pipelines contains:</p>


```
pipeline {
    agent none
    [...]    
Stages {
[...]
Stage('Build) {
    agent { label 'slave' }
    [...]
}
[...]
    }
}
```
<p>Much more space wasted on build nodes.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/rlindner81.jpg" title="rlindner81's avatar" /> rlindner81:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-42519?focusedId=434802&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-434802">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>I'm writing a short update since people, who google this problem inevitably end up here. In short, I agree with <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=stephenash" class="user-hover" rel="stephenash">stephenash</a> and disagree with <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=stephenconnolly" class="user-hover" rel="stephenconnolly">stephenconnolly</a>. Let me summarize the problem again from my point of view. Delegating the removal of dead PRs to the branch indexing causes Jenkins users to have the impossible choice of either</p>
<ul>
	<li>having frequent branch index runs, which will cost runtime resources for every PR / branch that gets built during the scan, which massively clogs up the build queue or</li>
	<li>having no or very rare branch index runs, which causes disk space to become smaller and smaller, since superfluous built data is not cleaned </li>
</ul>


<p>Needless to say both are bad.</p>

<p>I like the solution we use now, where we built a dedicated scripted pipeline "Delete Dead Branch Jobs" on Jenkins with the goal of doing the cleanup ourselves. So here is our code, very much without warranty of any kind, just as an inspiration to roll your own:</p>
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-groovy">#!/usr/bin/env groovy

<span class="code-keyword">import</span> org.jenkinsci.plugins.workflow.multibranch.BranchJobProperty;
<span class="code-keyword">import</span> jenkins.branch.Branch;
<span class="code-keyword">import</span> hudson.model.Job;

@NonCPS
List&lt;Job&gt; findDeadJobs() {
  <span class="code-keyword">def</span> jobs = Jenkins.instance.getItem(<span class="code-quote">"AFC"</span>).getAllJobs() <span class="code-comment">// &lt;-- change AFC to the name of your PR build pipeline
</span>  <span class="code-keyword">def</span> deadJobsSet = jobs.findAll({ it.getProperty(BranchJobProperty.class).getBranch() <span class="code-keyword">instanceof</span> Branch.Dead })
  <span class="code-keyword">return</span> <span class="code-keyword">new</span> ArrayList(deadJobsSet).sort({ it.getDisplayName() })
}

@NonCPS
void deleteJob(Job job) {
  job.delete()
  job.save()
}

node(<span class="code-quote">'master || built-<span class="code-keyword">in</span>'</span>) {
  <span class="code-keyword">def</span> deadJobs = findDeadJobs()
  println <span class="code-quote">"found ${deadJobs.size()} dead jobs"</span>
  deadJobs.each({
    println <span class="code-quote">"deleting dead job ${it.getDisplayName()}"</span>
    deleteJob(it)
  })
} </pre>
</div></div>
<p> </p>
</pre>
</details>

</ul>
</details>

<p>I'm writing a short update since people, who google this problem inevitably end up here. In short, I agree with <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=stephenash" class="user-hover" rel="stephenash">stephenash</a> and disagree with <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=stephenconnolly" class="user-hover" rel="stephenconnolly">stephenconnolly</a>. Let me summarize the problem again from my point of view. Delegating the removal of dead PRs to the branch indexing causes Jenkins users to have the impossible choice of either</p>
<ul>
	<li>having frequent branch index runs, which will cost runtime resources for every PR / branch that gets built during the scan, which massively clogs up the build queue or</li>
	<li>having no or very rare branch index runs, which causes disk space to become smaller and smaller, since superfluous built data is not cleaned </li>
</ul>


<p>Needless to say both are bad.</p>

<p>I like the solution we use now, where we built a dedicated scripted pipeline "Delete Dead Branch Jobs" on Jenkins with the goal of doing the cleanup ourselves. So here is our code, very much without warranty of any kind, just as an inspiration to roll your own:</p>

<pre>
#!/usr/bin/env groovy

<span class="code-keyword">import</span> org.jenkinsci.plugins.workflow.multibranch.BranchJobProperty;
<span class="code-keyword">import</span> jenkins.branch.Branch;
<span class="code-keyword">import</span> hudson.model.Job;

@​NonCPS
List<Job> findDeadJobs() {
  <span class="code-keyword">def</span> jobs = Jenkins.instance.getItem(<span class="code-quote">"AFC"</span>).getAllJobs() <span class="code-comment">// <-- change AFC to the name of your PR build pipeline
</span>  <span class="code-keyword">def</span> deadJobsSet = jobs.findAll({ it.getProperty(BranchJobProperty.class).getBranch() <span class="code-keyword">instanceof</span> Branch.Dead })
  <span class="code-keyword">return</span> <span class="code-keyword">new</span> ArrayList(deadJobsSet).sort({ it.getDisplayName() })
}

@​NonCPS
void deleteJob(Job job) {
  job.delete()
  job.save()
}

node(<span class="code-quote">'master || built-<span class="code-keyword">in</span>'</span>) {
  <span class="code-keyword">def</span> deadJobs = findDeadJobs()
  println <span class="code-quote">"found ${deadJobs.size()} dead jobs"</span>
  deadJobs.each({
    println <span class="code-quote">"deleting dead job ${it.getDisplayName()}"</span>
    deleteJob(it)
  })
} </pre>
<p> </p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/stephenconnolly.png" title="stephenconnolly's avatar" /> stephenconnolly:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-42519?focusedId=434807&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-434807">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=rlindner81" class="user-hover" rel="rlindner81">rlindner81</a> it is not entirely clear to me that you read this comment <a href="https://issues.jenkins.io/browse/JENKINS-42519?focusedCommentId=332368&amp;page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-332368" class="external-link" rel="nofollow">https://issues.jenkins.io/browse/JENKINS-42519?focusedCommentId=332368&amp;page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-332368</a></p>
<blockquote><p>The original design of the API was that pull requests would be handled as a separate source in the same multibranch... similarly you can add forks of the same repo... now in retrospect it one could argue that it was a bad design decision to do it that way in 2013... but there are users who use the multiple sources, so we cannot remove it... which means a branch being deleted in one source does not mean the branch job should be deleted.</p></blockquote>
<p>In other words there is a <b>technical</b> reason for the behaviour. If I had a time machine and travelled back to myself in 2013 I would tell that younger me that the separate source for pull requests would be a non-runner as that would use up too many API requests and that everyone would therefore just have a single source that returned both regular branches as well as pull requests, tags, etc... and in that case if you only have one source then yeah, dead branches can be deleted straight away.</p>

<p>So while you can join me in disagreeing with the technical design decision from 10 years ago, it is a fact that the system has been designed that way. You can not like the fact (I don't) but you cannot disagree that the system was designed that way. So you are misrepresenting my position when you say you disagree with me.</p>

<p>Anyway, it's all none of my business as I have moved on from working on Jenkins for a good while now... if somebody wanted to enhance the already complex code to say "oh I only have one source therefore I shall remove the dead branches straight away" I suspect some would be happy with that... but...</p>

<p>WARNING: If you can control when the branch project gets deleted, you can erase history of what that branch project did... so if you where to create a branch that downloaded and installed some stuff onto the agent that you shouldn't... well that could be visible in the logs... but then delete the branch and the job and it's logs are deleted immediately... then recreate the branch with innocent content and now any URLs from build notifications etc will take people to the new innocent job...</p>

<p>So if you did decide to add such a feature, be kind and allow people to turn it on or off according to their choosing as quite a few people may value the historical logs if even for a limited time</p>

<p>And as to the warning... there are multiple other ways to achieve similar effects without even needing to have a nefarious branch in the source code checked out... Evil dude could have a `go get` against a tag on a repo that evil dude controls... evil dude pushes an evil version to the tag, triggers the build, then deletes the tag and pushes a benign version as the tag... as long as nobody is tracking the `go.sum` file nobody after the fact will be able to reproduce the downloading of the bad code and it can do what it likes without outputting to the logs... perfectly hidden... that is of course why people either use lock files or use a dependency system where core infra has a write once contract (i.e. Maven Central)</p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=rlindner81" class="user-hover" rel="rlindner81">rlindner81</a> it is not entirely clear to me that you read this comment <a href="https://issues.jenkins.io/browse/JENKINS-42519?focusedCommentId=332368&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-332368" class="external-link" rel="nofollow">https://issues.jenkins.io/browse/JENKINS-42519?focusedCommentId=332368&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-332368</a></p>
<blockquote><p>The original design of the API was that pull requests would be handled as a separate source in the same multibranch... similarly you can add forks of the same repo... now in retrospect it one could argue that it was a bad design decision to do it that way in 2013... but there are users who use the multiple sources, so we cannot remove it... which means a branch being deleted in one source does not mean the branch job should be deleted.</p></blockquote>
<p>In other words there is a <b>technical</b> reason for the behaviour. If I had a time machine and travelled back to myself in 2013 I would tell that younger me that the separate source for pull requests would be a non-runner as that would use up too many API requests and that everyone would therefore just have a single source that returned both regular branches as well as pull requests, tags, etc... and in that case if you only have one source then yeah, dead branches can be deleted straight away.</p>

<p>So while you can join me in disagreeing with the technical design decision from 10 years ago, it is a fact that the system has been designed that way. You can not like the fact (I don't) but you cannot disagree that the system was designed that way. So you are misrepresenting my position when you say you disagree with me.</p>

<p>Anyway, it's all none of my business as I have moved on from working on Jenkins for a good while now... if somebody wanted to enhance the already complex code to say "oh I only have one source therefore I shall remove the dead branches straight away" I suspect some would be happy with that... but...</p>

<p>WARNING: If you can control when the branch project gets deleted, you can erase history of what that branch project did... so if you where to create a branch that downloaded and installed some stuff onto the agent that you shouldn't... well that could be visible in the logs... but then delete the branch and the job and it's logs are deleted immediately... then recreate the branch with innocent content and now any URLs from build notifications etc will take people to the new innocent job...</p>

<p>So if you did decide to add such a feature, be kind and allow people to turn it on or off according to their choosing as quite a few people may value the historical logs if even for a limited time</p>

<p>And as to the warning... there are multiple other ways to achieve similar effects without even needing to have a nefarious branch in the source code checked out... Evil dude could have a `go get` against a tag on a repo that evil dude controls... evil dude pushes an evil version to the tag, triggers the build, then deletes the tag and pushes a benign version as the tag... as long as nobody is tracking the `go.sum` file nobody after the fact will be able to reproduce the downloading of the bad code and it can do what it likes without outputting to the logs... perfectly hidden... that is of course why people either use lock files or use a dependency system where core infra has a write once contract (i.e. Maven Central)</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/rlindner81.jpg" title="rlindner81's avatar" /> rlindner81:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-42519?focusedId=434821&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-434821">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=stephenconnolly" class="user-hover" rel="stephenconnolly">stephenconnolly</a> I obviously misrepresented you and I'm genuinely sorry for that. What I meant to convey was, I disagree with the 2013 decision/architecture/software-design, in other words past-you, and not with present-you, who would do things differently. This problem has been bothering me every now and then for about 2 years. I expect it's much longer for you. I was somehow blindsided and didn't think to solve it like I did now. That's why I took the time to share my very hand-rolled solution, not with the intent of blaming, but helping others in my place to move forward.</p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=stephenconnolly" class="user-hover" rel="stephenconnolly">stephenconnolly</a> I obviously misrepresented you and I'm genuinely sorry for that. What I meant to convey was, I disagree with the 2013 decision/architecture/software-design, in other words past-you, and not with present-you, who would do things differently. This problem has been bothering me every now and then for about 2 years. I expect it's much longer for you. I was somehow blindsided and didn't think to solve it like I did now. That's why I took the time to share my very hand-rolled solution, not with the intent of blaming, but helping others in my place to move forward.</p>
